### PR TITLE
React

### PR DESCRIPTION
**Pull request by ItayGershman**

### CODE REVIEW

File: 5-Cookies/client-react/src/test/baseDriver.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { RenderResult } from "@testing-library/react";
+
+export abstract class BaseDriver {
+  protected wrapper?: RenderResult;
+
+  withWrapper(fn: (wrapper: RenderResult) => any) {
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/test/baseDriver.ts]: unused

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+describe("todo list item component", () => {
+  let driver : TodoListItemDriver;
+
+  beforeEach(() => {
+    driver = new TodoListItemDriver();
+  });
+
+  test("should display description text", async () => {
+    const item = { text: "test" };
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: generate strings with `chance`

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    driver.given.item(item);
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: return the driver itself and use chaining when possible
`driver.given.item(item).when.render()`

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    driver.when.render();
+
+    expect(driver.get.todoText()).toEqual(item.text);
+  });
+
+
+  test("should display switchMode status", async () => {
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: test description isn't very clear

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    driver.when.render();
+
+    expect(driver.get.todoStatus()).toEqual(false);
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx:25]: why false? at least describe the expectation of the product in the test description

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  });
+
+
+  test("should display editMode status", async () => {
+    driver.when.render();
+    
+    expect(driver.get.editModeStatus()).toBeFalsy();
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: be more specific `toEqual(false)` ?

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  });
+
+
+  test("should change editMode status to true", async () => {
+    driver.when.render(); 
+    driver.when.editButtonClick();
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: `clickEditButton`?

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+    expect(driver.get.editModeStatus()).toBeTruthy();
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: same here, `toEqual(true)` ?

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  });
+
+
+  test("should remove todo item from screen", () => {
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: try to describe the test from a product perspective (for instance: should remove an item when clicking the `trash` button)

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    const removeTodo: jest.Mock = jest.fn();
+    const item = mockTodo();
+    driver.given.item(item).removeTodo(removeTodo);
+
+    driver.when.render();
+    driver.when.removeButtonClick()
+
+    expect(removeTodo).toHaveBeenCalledWith(item.id);
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx:55]: this test is more of a unit test and in fact it doesn't match the description. i'd add another test that lists the todos and validate that it indeed isn't part of the list displayed

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx [hunk 11]
```diff
@@ -0,0 +1,53 @@
+import { mockTodo } from "../../../test/utils";
+import { TodoListItemDriver } from "./TodoListItem.driver";
+
+describe("todo list item component", () => {
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx:6]: i can think of missing tests here:
1. add a new todo
2. lists all todos
3. empty state (no todos?)
4. error state
5. many more 😅

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+import { render, fireEvent } from "@testing-library/react";
+import TodoListItem from "./TodoListItem";
+import { BaseDriver } from "../../../test/baseDriver";
+import dataHooks from "../../../dataHooks/dataHooks";
+import { mockTodo, throwErrorOnMissingWrapper } from "../../../test/utils";
+import { Item } from "../../../interfaces/interfaces";
+
+export class TodoListItemDriver extends BaseDriver {
+  private todoItem: Item = mockTodo();
+  private removeTodo: jest.Mock = jest.fn();
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: maybe `removeTodoMock`?

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+  private buttonClickHandler = (dataHook: string) => {
+    if (!this.wrapper) return throwErrorOnMissingWrapper();
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: throwing error doesn't require `return`

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    const buttonElem = this.wrapper.getByTestId(dataHook);
+    fireEvent.click(buttonElem);
+    return buttonElem;
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: why are you returning it? you wouldn't want to return the button item to the test but be more specific with the api you're exposing

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+  };
+
+  private getTodoText = () => {
+    return this.wrapper?.getByText(this.todoItem.text).innerHTML;
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: don't target dom by text, it's very fragile and can create confusion in the future (for instance what happens if the product changes and the text appears twice now?). use data hooks instead!

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+  };
+
+  private getTodoStatus = () => {
+    const switchElem = this.wrapper?.getByTestId(
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: there's inconsistency here - some times you're using `this.wrapper?` while other times `if (!this.wrapper)..` use one method everywhere (the latter 😄 )

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    return switchElem.checked;
+  };
+
+  private getEditModeStatus = () => {
+    if (!this.wrapper) return throwErrorOnMissingWrapper();
+    try {
+      if (this.wrapper.getByTestId(dataHooks.editTodoItem)) return true;
+    } catch (error) {
+      return false;
+    }
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: why `try..catch`? what are you expecting to catch here? if you're trying to check the existence of this element in the dom, there's probably a better way to do that

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+  };
+
+  given = {
+    item: (item: Partial<Item>) => {
+      this.todoItem = { ...this.todoItem, ...item };
+      return this.given;
+    },
+    removeTodo: (cb: jest.Mock) => {
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx:65]: not necessarily a `jest.Mock`, this could also be a standard `Function`

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+      return this.given;
+    },
+  };
+
+  when = {
+    render: () => {
+      this.wrapper = render(
+        <TodoListItem item={this.todoItem} removeTodo={this.removeTodo} />
+      );
+      return this.when;
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: go with `return this` everywhere instead. guess i've confused you here

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+const TodoListItem: FC<ITodoItem> = ({ item, removeTodo }) => {
+  const [isEditMode, setIsEditMode] = useState<boolean>(false);
+  const [todoItem, setTodoItem] = useState<Item>(item);
+
+  const dispatchEditTodo = async (newTodo: Item, id: Guid): Promise<void> => {
+    try {
+      await todosApi.editTodo(newTodo, id);
+    } catch (error) {
+      console.log(error);
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx]: why swallowing the error?

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+import { ITodoItem, TodoListItem } from "../../../interfaces/interfaces";
+import { classes } from "../../../js-styles/style";
+import IconButton from "../../general/IconButton";
+import Switch from "../../general/Switch";
+import EditTodoItem from "../EditTodoItem";
+import { todosApi } from "../../../../App";
+import dataHooks from "../../../dataHooks/dataHooks";
+import "../../../../style.css";
+import { Item } from "../../../interfaces/interfaces";
+import { Guid } from "../../../../../common/interfaces/Todo";
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx]: i suggest exposing everything from an index file, so consumers of the `common` types aren't aware of the internal structure

File: 5-Cookies/client-react/src/test/utils.ts [hunk 22]
```diff
@@ -0,0 +1,13 @@
+import { v4 as uuidv4 } from "uuid";
+
+export const mockTodo = () => {
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/test/utils.ts]: types!

Comment (tombenezra) [inline 5-Cookies/client-react/src/test/utils.ts]: change this function to be a builder of todos. so it'll accept overrides (same as you did in the driver) e.g something like:
```
export const buildTodoItem: Item = (overrides: Partial<item>) => ({
id: chance.guid(), 
text: chance.guid(), 
isFinished: false, 
...overrides 
})
```

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+      />
+      {isEditMode ? (
+        <EditTodoItem
+          item={todoItem}
+          editText={editText}
+          dataHook={dataHooks.editTodoItem}
+        />
+      ) : (
+        <span
+          className={`${classes.todoText} ${
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx]: [classnames](https://www.npmjs.com/package/classnames) to the rescue 🦸

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+    try {
+      await todosApi.editTodo(newTodo, id);
+    } catch (error) {
+      console.log(error);
+    }
+  };
+
+  const toggleTodo = async (): Promise<void> => {
+    setTodoItem({ ...todoItem, isFinished: !todoItem.isFinished });
+    dispatchEditTodo({ ...item, isFinished: !todoItem.isFinished }, item.id);
```
Comment (tombenezra) [inline 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx]: 1. you're not waiting for the async update in `dispatchEditTodo`
2. it might make more sense to update the server first (it is way more error prone), and update the UI afterwards if server has updated successfully. this way you won't get into broken state in the client

**Review by tombenezra — CHANGES_REQUESTED** [review]
This PR is huge and it's hard to follow. separate into mergable PRs when possible (AFAIK the server code was already reviewed and I assumed you're going to merge it)

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx (hunk 2 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: Alright

File: 5-Cookies/client-react/src/test/baseDriver.ts (hunk 1 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/test/baseDriver.ts]: removed

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx (hunk 2 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: Alright, I'll use that

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx (hunk 3 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: done

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx (hunk 4 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: changed to "should verify that a new item status is unfinished by default".
Actually, I'm not sure if this test is valuable, that do you think?

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx (hunk 6 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: yes, I changed to `toEqual(false)`

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx (hunk 7 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: changed to clickOnEditButton

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx (hunk 8 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: changed to `toEqual(true)`

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx (hunk 9 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx]: Alright, done that, and for future tests I will come up with better descriptions

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx (hunk 10 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx:55]: I agree, but this component doesn't render the todos, so I will develop this test on the list component

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx (hunk 11 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx:6]: These tests are for the entire application, I will get to that in the next few days.

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx (hunk 12 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: Agree

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx (hunk 14 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: Yes, my bad, I don't need to return it

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx (hunk 15 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: I saw that in the library and thought that it was ok to use that, now I understand that it isn't, so I'll change it to data hooks

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx (hunk 16 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: I let the autocomplete to control my coding style... I'll change that 😅

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx (hunk 17 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: This is what I'm trying to do and didn't find any better way to do it, do you have a suggestion on how to find the existence of an element?

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx (hunk 18 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx:65]: I thought that on tests we only pass a mock function, do we sometimes create regular functions?

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx (hunk 19 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.driver.tsx]: done

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx (hunk 20 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx]: Actually, I just changed that to toast an error with `react-toastify` library to alert the user of a problem.

File: 5-Cookies/client-react/src/test/utils.ts (hunk 22 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/test/utils.ts]: OOPS, fixed it

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx (hunk 21 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx]: Do you mean expose every **interface** of the app from an index.file?

File: 5-Cookies/client-react/src/test/utils.ts (hunk 22 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/test/utils.ts]: Ok, I'll will change `mockTodo` function to your suggestion

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx (hunk 24 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx]: Alright, I understand your intent, I will follow it

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx (hunk 23 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.tsx]: Now it looks much nicer!

File: 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx (hunk 5 above)
Comment (ItayGershman) [reply 5-Cookies/client-react/src/components/TodoList/TodoListItem/TodoListItem.spec.tsx:25]: this test has been removed
